Messaging/EventMessagingHandler.cs

```diff
@@ -0,0 +1,229 @@
+        private void ProcessPacket(Packet packet)
+        {
+            //Packets we handled in the UX.
+            switch (packet.Type)
+            {
+                case PacketType.Debug:
+                    var debug = (DebugPacket)packet;
+                    OnDebugEvent(debug);
+                    break;
+
+                case PacketType.Log:
+                    var log = (LogPacket)packet;
+                    OnLogEvent(log);
+                    break;
+
+                case PacketType.RuntimeError:
+                    var runtime = (RuntimeErrorPacket)packet;
+                    OnRuntimeErrorEvent(runtime);
+                    break;
+
+                case PacketType.HandledError:
+                    var handled = (HandledErrorPacket)packet;
+                    OnHandledErrorEvent(handled);
+                    break;
+
+                case PacketType.BacktestResult:
+                    var result = (BacktestResultPacket)packet;
+                    OnBacktestResultEvent(result);
+                    break;
+            }
+
+            if (StreamingApi.IsEnabled)
+            {
+                StreamingApi.Transmit(_job.UserId, _job.Channel, packet);
+            }
+        }
+
+        /// <summary>
+        /// Raise a debug event safely
+        /// </summary>
+        protected virtual void OnDebugEvent(DebugPacket packet)
+        {
+            if (DebugEvent != null)
+            {
+                DebugEvent(packet);
+            }
+        }
+
+        /// <summary>
+        /// Handler for consumer ready code.
+        /// </summary>
+        public virtual void OnConsumerReadyEvent()
+        {
+            if (ConsumerReadyEvent != null)
+            {
+                ConsumerReadyEvent();
+            }
+        }
+
+        /// <summary>
+        /// Raise a log event safely
+        /// </summary>
+        protected virtual void OnLogEvent(LogPacket packet)
+        {
+            if (LogEvent != null)
+            {
+                LogEvent(packet);
+            }
+        }
+
+        /// <summary>
+        /// Raise a handled error event safely
+        /// </summary>
+        protected virtual void OnHandledErrorEvent(HandledErrorPacket packet)
+        {
+            if (HandledErrorEvent != null)
+            {
+                HandledErrorEvent(packet);
+            }
+        }
+
+        /// <summary>
+        /// Raise runtime error safely
+        /// </summary>
+        protected virtual void OnRuntimeErrorEvent(RuntimeErrorPacket packet)
+        {
+            if (RuntimeErrorEvent != null)
+            {
+                RuntimeErrorEvent(packet);
+            }
+        }
+
+        /// <summary>
+        /// Raise a backtest result event safely.
+        /// </summary>
+        protected virtual void OnBacktestResultEvent(BacktestResultPacket packet)
+        {
+            if (BacktestResultEvent != null)
+            {
+                BacktestResultEvent(packet);
```

Comment:
It's a threading concern to not assign the event handler to a local variable, the event handler may become null between the null check and the invocation

http://stackoverflow.com/questions/5147397/why-assign-a-handler-to-an-event-before-calling-it